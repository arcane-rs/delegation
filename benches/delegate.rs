use criterion::{criterion_group, criterion_main, Criterion};
use delegation::delegate;

#[delegate]
trait AsStr {
    fn as_str(&self) -> &str;
}

impl AsStr for String {
    fn as_str(&self) -> &str {
        self
    }
}

macro_rules! def_enum {
    ($($variant:ident),+) => {
        #[delegate(derive(AsStr))]
        enum Name {
            $($variant(String)),+
        }
    };
}

macro_rules! def_bench {
    ($($variant:ident),*) => {
        fn bench(bencher: &mut ::criterion::Bencher) {
            $(
                bencher.iter(|| {
                    let name = ::criterion::black_box(
                        Name::$variant(String::from("Test"))
                    );

                    assert_eq!(name.as_str(), "Test");
                });
            )*
        }
    };
}

macro_rules! define {
    ($($variant:ident),*) => {
        def_enum!($($variant),*);
        def_bench!($($variant),*);
    };
}

define!(
    A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y,
    Z
);

fn delegate_benchmark(c: &mut Criterion) {
    c.bench_function("delegate", bench);
}

criterion_group!(benches, delegate_benchmark);
criterion_main!(benches);
