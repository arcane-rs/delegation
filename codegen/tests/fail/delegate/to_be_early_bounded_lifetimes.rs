use delegation::delegate;

#[delegate]
trait PrependWith {
    fn prepend_with<'s>(&self, prefix: &'s str) -> String;
}

impl PrependWith for String {
    fn prepend_with<'s>(&self, prefix: &'s str) -> String {
        format!("{prefix}{self}")
    }
}

#[delegate(derive(PrependWith))]
enum EitherDef {
    Left(String),
    Right(String),
}

fn main() {}
