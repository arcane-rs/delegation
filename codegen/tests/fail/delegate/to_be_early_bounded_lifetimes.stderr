error: lifetime `'s` are limited to be early-bounded. Consider adding `'s: 's` bound or replace them with `'_'. See `rust-lang/rust#87803` for details.
 --> tests/fail/delegate/to_be_early_bounded_lifetimes.rs:5:5
  |
5 |     fn prepend_with<'s>(&self, prefix: &'s str) -> String;
  |     ^^

error: cannot find macro `PrependWith` in this scope
  --> tests/fail/delegate/to_be_early_bounded_lifetimes.rs:14:19
   |
14 | #[delegate(derive(PrependWith))]
   |                   ^^^^^^^^^^^

error[E0405]: cannot find trait `PrependWith` in this scope
 --> tests/fail/delegate/to_be_early_bounded_lifetimes.rs:8:6
  |
8 | impl PrependWith for String {
  |      ^^^^^^^^^^^ not found in this scope
